Source/ASMultiplexImageNode.mm

```diff
@@ -835,30 +846,67 @@ - (void)_downloadImageWithIdentifier:(id)imageIdentifier URL:(NSURL *)imageURL c
-                                                          //Getting a result back for a different download identifier, download must not have been successfully canceled
-                                                          if (ASObjectIsEqual(_downloadIdentifier, downloadIdentifier) == NO && downloadIdentifier != nil) {
-                                                            return;
-                                                          }
-                                                          
-                                                          completionBlock([imageContainer asdk_image], error);
-                                                          
-                                                          // Delegateify.
-                                                          if (strongSelf->_delegateFlags.downloadFinish)
-                                                            [strongSelf->_delegate multiplexImageNode:weakSelf didFinishDownloadingImageWithIdentifier:imageIdentifier error:error];
-                                                        }]];
-    [self _updateProgressImageBlockOnDownloaderIfNeeded];
+    __typeof__(self) strongSelf = weakSelf;
+    if (!strongSelf)
+      return;
+
+    dispatch_queue_t callbackQueue = dispatch_get_main_queue();
+
+    id downloadIdentifier;
+    if (strongSelf->_downloaderFlags.downloaderImplementsDownloadWithPriority
+        && strongSelf->_isInImageDownloaderPriorityExperiment) {
+
+      /*
+       Decide a priority based on the current interface state of this node.
+       It can happen that this method was called when the node entered preload state
+       but the interface state, at this point, tells us that the node is (going to be) visible,
+       If that's the case, we jump to a higher priority directly.
+       */
+      ASImageDownloaderPriority priority = ASImageDownloaderPriorityWithInterfaceState(strongSelf.interfaceState);
+      downloadIdentifier = [strongSelf->_downloader downloadImageWithURL:imageURL
+                                                                priority:priority
+                                                           callbackQueue:callbackQueue
+                                                        downloadProgress:downloadProgressBlock
+                                                              completion:completion];
+    } else {
+      /*
+       Kick off a download with default priority.
+       The actual "default" value is decided by the downloader.
+       ASBasicImageDownloader and ASPINRemoteImageDownloader both use ASImageDownloaderPriorityImminent
+       which is mapped to NSURLSessionTaskPriorityDefault.
+
+       This means that preload and display nodes use the same priority
+       and their requests are put into the same pool.
+       */
+      downloadIdentifier = [strongSelf->_downloader downloadImageWithURL:imageURL
```

Comment:
can we call this we a priority also so that it do not need to be called differently in if-else?

Reply:
I'd like to let the concrete downloader to decide what the default priority is. Right now both `ASBasicImageDownloader` and `ASPINRemoteImageDownloader` have the same default value, but other custom ones may have a different one.